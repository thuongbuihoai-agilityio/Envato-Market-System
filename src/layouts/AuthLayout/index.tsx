import areEqual from 'react-fast-compare';
import { Box, Flex } from '@chakra-ui/react';
import { ReactNode, memo, useMemo } from 'react';

// Constants
import { IMAGES, TITLES } from '@app/constants';

// HOCs
import { withAuthentication, withErrorBoundary } from '@app/hocs';

// Components
import { Benefit, Divider, Logo, SwitchTheme } from '@app/components';
import AuthHeading from './Heading';
import AuthFooter from './Footer';

// Types
import { TImageDetails } from '@app/interfaces';

type TAuthLayoutProps = {
  children?: ReactNode;
  title?: string;
  isSignInForm?: boolean;
};

const AuthLayoutComponent = ({
  children,
  isSignInForm = true,
}: TAuthLayoutProps): JSX.Element => {
  const title: string = useMemo(
    (): string => (isSignInForm ? TITLES.SIGN_IN : TITLES.SIGN_UP),
    [isSignInForm],
  );
  const { url, alt, width, height }: TImageDetails = useMemo(
    (): TImageDetails => (isSignInForm ? IMAGES.SIGN_IN : IMAGES.SIGN_UP),
    [isSignInForm],
  );

  return (
    <Flex width="100%" minH="100vh">
      <Box
        as="section"
        p="40px 0 48px"
        flex={1}
        w={{
          base: '100%',
          md: 'unset',
        }}
        bg="background.body.secondary"
      >
        <Flex justifyContent="space-between" px={12}>
          <Logo />
          <SwitchTheme />
        </Flex>
        <Box
          w={{
            base: '100%',
            sm: 425,
            md: 460,
          }}
          margin="auto"
          pt={24}
          pb={16}
          px={5}
          sx={{
            boxSizing: {
              base: 'border-box',
              md: 'unset',
            },
          }}
        >
          <AuthHeading title={title} />
          <Divider content={TITLES.AUTH_DiVIDER} />
          {children}
          <AuthFooter />
        </Box>
      </Box>
      <Benefit
        image={{ url: url, width: width, height: height }}
        alt={alt}
        heading="Speady, Easy and Fast"
      />
    </Flex>
  );
};
const AuthLayout = memo(
  withErrorBoundary(withAuthentication(AuthLayoutComponent)),
  areEqual,
);

export default AuthLayout;
