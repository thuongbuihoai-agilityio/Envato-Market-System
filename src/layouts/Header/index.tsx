import { memo } from 'react';

// Components
import {
  Box,
  Flex,
  Heading,
  Text,
  theme,
  useColorModeValue,
} from '@chakra-ui/react';
import { Dropdown, IconButton, Logo, SwitchTheme } from '@app/components';

// Assets
import { Bell, Email, Gift } from '@app/components/Icons';

// hooks
import { authStore } from '@app/stores';

interface HeaderProps {
  name?: string;
}

const HeaderComponent = ({ name }: HeaderProps) => {
  const colorFill = useColorModeValue(
    theme.colors.gray[800],
    theme.colors.white,
  );

  const username = authStore(
    ({ user }): string | undefined => `${user?.firstName} ${user?.lastName}`,
  );
  const avatarURL = authStore(
    (state): string | undefined => state.user?.avatarURL,
  );

  return (
    <Flex
      h="100%"
      maxW="full"
      bg="background.component.primary"
      alignItems="start"
      px={{ base: 6, xl: 10 }}
      py={6}
      justifyContent="space-between"
      direction={{
        base: 'column',
        default: 'row',
      }}
    >
      <Flex
        minW="full"
        display={{ base: 'inline-flex', md: 'none' }}
        justifyContent="space-between"
        alignItems="center"
      >
        <Logo />
        <Flex display={{ base: 'flex' }} height="min-content" gap={4}>
          <Flex
            display={{ base: 'none', default: 'inline-flex' }}
            justifyContent="end"
            w="fit-content"
            gap={3}
          >
            <SwitchTheme />

            <IconButton isNotification>
              <Bell color={colorFill} />
            </IconButton>

            <IconButton isNotification hasNewNotification>
              <Email color={colorFill} />
            </IconButton>

            <IconButton isNotification>
              <Gift color={colorFill} />
            </IconButton>
          </Flex>
          <Dropdown name="John Doe" permission="Super Admin" src={avatarURL} />
        </Flex>
      </Flex>
      <Flex
        display={{ base: 'inline-flex', default: 'none' }}
        alignSelf="end"
        w="100%"
        gap={3}
        pt={2.5}
        justifyContent={{
          base: 'space-between',
          '0.8sm': 'end',
        }}
      >
        <SwitchTheme />

        <IconButton isNotification>
          <Bell color={colorFill} />
        </IconButton>

        <IconButton isNotification hasNewNotification>
          <Email color={colorFill} />
        </IconButton>

        <IconButton isNotification>
          <Gift color={colorFill} />
        </IconButton>
      </Flex>
      <Box display={{ base: 'none', md: 'inline' }} minW={185}>
        <Heading
          as="h1"
          fontSize="3xl"
          fontFamily="primary"
          fontWeight="bold"
          color="text.primary"
        >
          {name}
        </Heading>
        <Text fontSize="sm" color="text.secondary" fontWeight="medium">
          Letâ€™s check your update today
        </Text>
      </Box>

      <Flex
        w={{ base: '100%', xl: 407, '3xl': 530 }}
        justifyContent={{ base: 'end', xl: 'space-between' }}
        direction={{
          base: 'column',
          xl: 'row',
        }}
        gap={4}
        pl={{
          base: 0,
          md: 5,
          lg: 24,
          xl: 0,
        }}
      >
        <Flex
          gap={43}
          alignSelf={{
            base: 'end',
            xl: 'baseline',
          }}
        >
          <Flex
            display={{ base: 'none', md: 'inline-flex' }}
            borderRight="1px"
            borderColor="border.primary"
            height="min-content"
            pr={43}
            minW={310}
            justifyContent="space-between"
          >
            <SwitchTheme />

            <IconButton isNotification>
              <Bell color={colorFill} />
            </IconButton>

            <IconButton isNotification hasNewNotification>
              <Email color={colorFill} />
            </IconButton>

            <IconButton isNotification>
              <Gift color={colorFill} />
            </IconButton>
          </Flex>

          <Box display={{ base: 'none', md: 'inline-flex' }}>
            <Dropdown
              name={username}
              permission="Super Admin"
              src={avatarURL}
            />
          </Box>
        </Flex>
      </Flex>
    </Flex>
  );
};

const Header = memo(HeaderComponent);

export default Header;
