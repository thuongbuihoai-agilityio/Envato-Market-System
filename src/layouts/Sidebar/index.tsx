import { memo, useMemo } from 'react';
import isEqual from 'react-fast-compare';

// components
import { ExpandSidebar, MiniSidebar } from '@app/components';
import { TMenuItem } from '@app/components/common/Menu';

// Constants
import {
  HELP_ITEM_LIST,
  AUTHENTICATION_ROLE,
  MENU_ITEM_LIST,
} from '@app/constants';

// Stores
import { authStore } from '@app/stores';

// Interfaces
import { TUserDetail } from '@app/interfaces';

export type SidebarProps = {
  menuItem?: TMenuItem[];
  roleAdmin?: string;
  onClose: () => void;
  onOpen: () => void;
  isOpen: boolean;
};

const Sidebar = ({ onClose, onOpen, isOpen }: SidebarProps) => {
  const { user } = authStore();

  const { role = AUTHENTICATION_ROLE.MEMBER } = user as TUserDetail;
  const menuItem = useMemo(() => MENU_ITEM_LIST(role), [role]);

  return (
    <>
      <ExpandSidebar
        roleAdmin={role}
        menuItem={[...(menuItem as TMenuItem[]), ...HELP_ITEM_LIST]}
        onClose={onClose}
        onOpen={onOpen}
        isOpen={!isOpen}
      />
      <MiniSidebar
        roleAdmin={role}
        menuItem={[...(menuItem as TMenuItem[]), ...HELP_ITEM_LIST]}
        onClose={onClose}
        isOpen={isOpen}
      />
    </>
  );
};

const SideBarComponent = memo(Sidebar, isEqual);

export default SideBarComponent;
