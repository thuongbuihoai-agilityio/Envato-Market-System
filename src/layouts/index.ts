export { default as MainLayout } from './MainLayout';
export { default as AuthLayout } from './AuthLayout';
export { default as SideBar } from './Sidebar';
export { default as Header } from './Header';
