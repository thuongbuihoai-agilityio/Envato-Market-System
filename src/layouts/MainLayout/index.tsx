import { Outlet, useLocation } from 'react-router-dom';
import { useCallback, useEffect } from 'react';
import {
  Box,
  Flex,
  Image,
  useDisclosure,
  useToast,
  useMediaQuery,
} from '@chakra-ui/react';
import { SubmitHandler, useForm } from 'react-hook-form';

// Constants
import {
  ERROR_MESSAGES,
  IMAGES,
  SHOW_TIME,
  SIDEBAR,
  SUCCESS_MESSAGES,
  TITLES_HEADER,
} from '@app/constants';

// Components
import { Lazy, Modal, PinCode } from '@app/components';

// Layouts
import { Header, SideBar } from '@app/layouts';

// HOCs
import { withErrorBoundary } from '@app/hocs/withErrorBoundary';

// Stores
import { TAuthStoreData, authStore } from '@app/stores';

// Hooks
import { useAuth, usePinCode } from '@app/hooks';

export type TPinCodeForm = {
  pinCode: string;
  userId: string;
};

const Layout = () => {
  const { isOpen, onOpen, onClose } = useDisclosure({
    defaultIsOpen: true,
  });

  const [isTablet] = useMediaQuery('(min-width: 768px) and (max-width: 992px)');

  const location = useLocation();

  const { pathname } = location;

  // Open mini sidebar on tablet
  useEffect(() => {
    if (isTablet) {
      onOpen();
    }
  }, [isTablet, onOpen]);

  const {
    isOpen: isPinCodeModalOpen,
    onClose: onClosePinCodeModal,
    onOpen: onOpenPinCodeModal,
  } = useDisclosure();

  const {
    control,
    handleSubmit,
    formState: { isValid, isSubmitting },
  } = useForm<TPinCodeForm>({
    defaultValues: {
      pinCode: '',
    },
  });

  const user = authStore((state): TAuthStoreData['user'] => state.user);

  const { setUser } = useAuth();

  const toast = useToast();

  const { handleSetPinCode } = usePinCode();

  useEffect(() => {
    if (!user?.pinCode) {
      onOpenPinCodeModal();
    }
  }, []);

  const onSubmitPinCode: SubmitHandler<TPinCodeForm> = useCallback(
    (data) => {
      if (user) {
        data.userId = user.id;

        try {
          handleSetPinCode(data);

          setUser({ user: { ...user, pinCode: data.pinCode } });

          onClosePinCodeModal();

          toast({
            title: SUCCESS_MESSAGES.SET_PIN_CODE.title,
            description: SUCCESS_MESSAGES.SET_PIN_CODE.description,
            status: 'success',
            duration: SHOW_TIME,
            isClosable: true,
            position: 'top-right',
          });
        } catch (error) {
          toast({
            title: ERROR_MESSAGES.SET_PIN_CODE.title,
            description: ERROR_MESSAGES.SET_PIN_CODE.description,
            status: 'error',
            duration: SHOW_TIME,
            isClosable: true,
            position: 'top-right',
          });
        }
      }
    },
    [handleSetPinCode, onClosePinCodeModal, setUser, toast, user],
  );

  const renderPinCodeModalBody = useCallback(
    () => (
      <PinCode
        control={control}
        isDisabled={!isValid || isSubmitting}
        onSubmit={handleSubmit(onSubmitPinCode)}
        onClose={onClosePinCodeModal}
      />
    ),
    [
      control,
      handleSubmit,
      isSubmitting,
      isValid,
      onClosePinCodeModal,
      onSubmitPinCode,
    ],
  );

  return (
    <>
      <Flex w="full" h="full" bg="background.body.primary">
        <Box
          pl={{
            base: 0,
            md: !isOpen
              ? SIDEBAR.EXPAND_SIDEBAR_WIDTH
              : SIDEBAR.MINI_SIDEBAR_WIDTH,
            // lg: !isOpen
            //   ? SIDEBAR.EXPAND_SIDEBAR_WIDTH
            //   : SIDEBAR.MINI_SIDEBAR_WIDTH,
            // '4xl': !isOpen
            //   ? SIDEBAR.EXPAND_SIDEBAR_WIDTH
            //   : SIDEBAR.MINI_SIDEBAR_WIDTH,
          }}
          w="full"
          minH="100vh"
          h="full"
          sx={{
            transition: 'all .25s ease-in-out',
          }}
        >
          <SideBar isOpen={isOpen} onOpen={onOpen} onClose={onClose} />

          <Header
            name={
              TITLES_HEADER[`${pathname.slice(1)}`] || TITLES_HEADER.DEFAULT
            }
          />

          {/* Button to show Sidebar on mobile */}
          <Image
            src={IMAGES.LEFT_ARROW.url}
            alt={IMAGES.LEFT_ARROW.alt}
            fallbackSrc={IMAGES.FALLBACK.url}
            fallbackStrategy="onError"
            position="fixed"
            w={4}
            h={10}
            top={8}
            transform="rotate(180deg)"
            left={0}
            onClick={onClose}
          />

          {/* Content of the page */}
          <Lazy>
            <Outlet />
          </Lazy>
        </Box>
      </Flex>
      <Modal
        title="Please set the PIN code to your account"
        isOpen={isPinCodeModalOpen}
        onClose={onClosePinCodeModal}
        renderBody={renderPinCodeModalBody}
      />
    </>
  );
};

const MainLayout = withErrorBoundary(Layout);

export default MainLayout;
