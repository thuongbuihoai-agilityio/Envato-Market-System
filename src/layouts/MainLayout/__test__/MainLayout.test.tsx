import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { MemoryRouter } from 'react-router-dom';

import { MainLayout } from '@app/layouts';

describe('MainLayout Component', () => {
  it('Should render match with snapshot.', async () => {
    const { container } = render(<MainLayout />, { wrapper: MemoryRouter });

    expect(container).toMatchSnapshot();
  });
});
