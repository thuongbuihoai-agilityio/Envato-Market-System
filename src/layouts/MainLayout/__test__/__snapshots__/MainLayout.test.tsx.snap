// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainLayout Component Should render match with snapshot. 1`] = `
<div>
  <div
    class="chakra-stack css-1pkffy6"
  >
    <p
      class="chakra-text css-oz7p8m"
      data-testid="error-title"
    >
      An error has been occurred!!
    </p>
    <p
      class="chakra-text css-1ly9ogx"
      data-testid="error-message"
    >
      win.matchMedia is not a function
    </p>
  </div>
</div>
`;
